core/src/test/java/feast/core/grpc/CoreServiceImplTest.java

```diff
@@ -0,0 +1,135 @@
+            .build()
+            .start());
+    CoreServiceBlockingStub blockingStub =
+        CoreServiceGrpc.newBlockingStub(
+            grpcCleanup.register(
+                InProcessChannelBuilder.forName(serverName).directExecutor().build()));
+
+    GetUploadUrlResponse response =
+        blockingStub.getUploadUrl(
+            GetUploadUrlRequest.newBuilder().setFileType(FileType.CSV).build());
+
+    // Make sure the signed URL response allows upload using PUT method and the link is valid for
+    // the next 5 minutes
+    assertEquals(response.getHttpMethod(), HttpMethod.PUT);
+    assertTrue(response.getExpiration().getSeconds() > System.currentTimeMillis() / 1000);
+    assertTrue(
+        response.getExpiration().getSeconds() < (System.currentTimeMillis() + 6 * 60_000) / 1000);
+
+    // Test that we can upload data using the signed URL from the response
+    String uploadUrl = response.getUrl();
+    URL url = new URL(uploadUrl);
+    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+    conn.setDoOutput(true);
+    conn.setRequestMethod("PUT");
+    OutputStreamWriter out = new OutputStreamWriter(conn.getOutputStream());
+    out.write("1,2,3\n4,5,6");
+    out.close();
+    conn.getInputStream();
+
+    // Ensure the data is uploaded correctly
+    Storage storage = StorageOptions.getDefaultInstance().getService();
+    String fileName =
+        uploadUrl.substring(
+            "https://storage.googleapis.com/feast-templocation-kf-feast/".length(),
+            uploadUrl.indexOf("?"));
+    File tempFile = File.createTempFile("prefix", "suffix");
+    tempFile.deleteOnExit();
+    storage.get(BlobId.of(bucketName, fileName)).downloadTo(tempFile.toPath());
+    List<String> strings = Files.readAllLines(tempFile.toPath());
+    assertEquals(strings.get(0), "1,2,3");
+    assertEquals(strings.get(1), "4,5,6");
+  }
+
+  @Test
+  public void getBucketNameFromWorkspace() {
+    List<Object[]> testCases =
+        Arrays.asList(
+            new Object[] {"gs://bucket", "bucket"},
+            new Object[] {"gs://bucket/", "bucket"},
+            new Object[] {"gs://bucket/dir", "bucket"},
+            new Object[] {null, new IllegalArgumentException()},
+            new Object[] {"", new IllegalArgumentException()},
+            new Object[] {"/", new IllegalArgumentException()},
+            new Object[] {"/local/dir", new IllegalArgumentException()},
+            new Object[] {"gs", new IllegalArgumentException()},
+            new Object[] {"gs", new IllegalArgumentException()},
+            new Object[] {"gs://", new IllegalArgumentException()});
+
+    for (Object[] testCase : testCases) {
+      String input = (String) testCase[0];
+      Object expected = testCase[1];
+
+      try {
+        assertEquals(expected, CoreServiceImpl.getBucketNameFromWorkspace(input));
+      } catch (IllegalArgumentException e) {
+        if (!(expected instanceof IllegalArgumentException)) {
```

Comment:
What happens here if `expected instanceof IllegalArgumentException` is true?

Reply:
Will the code arrive at this block?

Reply:
this block
```
catch (IllegalArgumentException e) {
        if (!(expected instanceof IllegalArgumentException)) {
          fail(
              String.format(
                  "Exception thrown for input '%s', expected output: '%s'", input, expected));
        }
}
```
is actually meant to **test exceptions being thrown**.

In JUnit4 the usual way to test exceptions being thrown, is to create another test method with expected annotation like so
```
@Test(expected = IndexOutOfBoundsException.class)
public void testIndexOutOfBoundsException() {
    ArrayList emptyList = new ArrayList();
    Object o = emptyList.get(0);
}
```

but I feel putting all the test cases in one place makes it easier to see the behaviour of the _method under test_.

The code will arrive at this block, if the _method under test_ **throws an exception when we do not expect it to**, i.e. if we implement `CoreServiceImpl.getBucketNameFromWorkspace` incorrectly. Instead of returning "bucket" when argument is "gs://bucket", the method throws `IllegalArgumentException`

Let me know if you think this test can be made clearer.